### ISSUE TITLE
Feature request: extend API Gateway schemas with custom schema in `body` rather than `string`

### ISSUE THREAD

**Issue by julbrs** [aws-powertools/powertools-lambda-typescript#2844]
### Expected Behavior

I expect the parsing event document to work: https://docs.powertools.aws.dev/lambda/typescript/latest/utilities/parser/#parse-events

### Current Behavior

When I implement such code, I ended with the error:

```
ParseError: Failed to parse schema.
```

Because no `envelope` was set.

### Code snippet

My failing code:

```
import { newNetsuiteUserSchema } from '@/packages/core/model/netsuite';

import middy from '@middy/core';
import { parser } from '@aws-lambda-powertools/parser/middleware';
import { APIGatewayProxyEventV2Schema } from '@aws-lambda-powertools/parser/schemas';
import { z } from 'zod';

export const baseHandler = async (event: z.infer<typeof newNetsuiteUserSchema>) => {
  return {
    statusCode: 200,
    body: JSON.stringify(event),
  };
};

export const handler = middy(baseHandler).use(parser({ schema: newNetsuiteUserSchema }));
```

with: 

```
export const newNetsuiteUserSchema = z.object({
  firstName: z.string().min(1),
  lastName: z.string().min(1),
  email: z.string().email().min(1),
  role: z.nativeEnum(UserRoles),
  custentity_lba_customer_center_access: z.string().min(1),
});
```

### Steps to Reproduce

I am using SST (sst.dev) to build my project.

### Possible Solution

I think it's mandatory to specify the envelope like that:

```
import { newNetsuiteUserSchema } from '@/packages/core/model/netsuite';

import middy from '@middy/core';
import { parser } from '@aws-lambda-powertools/parser/middleware';
import { APIGatewayProxyEventV2Schema } from '@aws-lambda-powertools/parser/schemas';
import { z } from 'zod';
import { ApiGatewayV2Envelope } from '@aws-lambda-powertools/parser/envelopes';

export const baseHandler = async (event: z.infer<typeof newNetsuiteUserSchema>) => {
  return {
    statusCode: 200,
    body: JSON.stringify(event),
  };
};

export const handler = middy(baseHandler).use(
  parser({ schema: newNetsuiteUserSchema, envelope: ApiGatewayV2Envelope })
);
```

### Powertools for AWS Lambda (TypeScript) version

latest (2.5.0-beta)

### AWS Lambda function runtime

20.x

### Packaging format used

npm

### Execution logs

```Shell
},
     {
       "expected": "'Admin' | 'User'",
       "received": "undefined",
       "code": "invalid_type",
       "path": [
         "role"
       ],
       "message": "Required"
     },
     {
       "code": "invalid_type",
       "expected": "string",
       "received": "undefined",
       "path": [
         "custentity_lba_customer_center_access"
       ],
       "message": "Required"
     }
   ].
       at parse (/Users/julien/work/lab/customer-center/node_modules/.pnpm/@aws-lambda-powertools+parser@2.5.0-beta_@middy+core@5.4.3_zod@3.23.8/node_modules/@aws-lambda-powertools/parser/lib/esm/parser.js:41:15)
       at before (/Users/julien/work/lab/customer-center/node_modules/.pnpm/@aws-lambda-powertools+parser@2.5.0-beta_@middy+core@5.4.3_zod@3.23.8/node_modules/@aws-lambda-powertools/parser/lib/esm/middleware/parser.js:33:25)
       at runMiddlewares (/Users/julien/work/lab/customer-center/node_modules/.pnpm/@middy+core@5.4.3/node_modules/@middy/core/index.js:230:23)
       at runRequest (/Users/julien/work/lab/customer-center/node_modules/.pnpm/@middy+core@5.4.3/node_modules/@middy/core/index.js:154:11)
       at middyHandler (/Users/julien/work/lab/customer-center/node_modules/.pnpm/@middy+core@5.4.3/node_modules/@middy/core/index.js:43:12)
       at file:///Users/julien/work/lab/customer-center/node_modules/.pnpm/sst@2.43.4_@aws-sdk+client-sso-oidc@3.606.0_@aws-sdk+client-sts@3.606.0_aws-crt@1.21.3__aws-c_gdj3g64c23iicrgcjjtaychpzi/node_modules/sst/support/nodejs-runtime/index.mjs:148:22
       at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
```

**Comment by dreamorosi** [comment]
Hi @julbrs, thank you for opening this issue.

Based on the description and logs, it seems that the utility is working as expected.

When you specify a schema, the event will be validated using that schema as-is. In your case, based on your schema, the event that would allow your handler to pass validation would be something like:

```json
{
  "firstName": "foo",
  "lastName": "bar",
  "email": foo.bar@example.com",
  "role": "guest",
  "custentity_lba_customer_center_access": "baq",
}
```

In your example however, you included the `APIGatewayProxyEventV2Schema`, which suggests you're using Amazon API Gateway with AWS Lambda proxy integration.

If this is the case, the event will look something like this:

```json
{
  "version": "2.0",
  "routeKey": "POST /lambda",
  "rawPath": "/lambda",
  "rawQueryString": "",
  "headers": {
    "accept": "*/*",
    "accept-encoding": "gzip, deflate",
    "authorization": "Bearer foo",
    "content-length": "0",
    "host": "lsw1ro4ipb.execute-api.eu-west-1.amazonaws.com",
    "user-agent": "HTTPie/3.2.2",
    "x-amzn-trace-id": "Root=1-66705bc7-2b4257df30cbee696ef2cf28",
    "x-forwarded-for": "15.248.3.126",
    "x-forwarded-port": "443",
    "x-forwarded-proto": "https"
  },
  "requestContext": {
    "accountId": "123456789012",
    "apiId": "lsw1ro4ipb",
    "authorizer": {
      "lambda": null
    },
    "domainName": "lsw1ro4ipb.execute-api.eu-west-1.amazonaws.com",
    "domainPrefix": "lsw1ro4ipb",
    "http": {
      "method": "POST",
      "path": "/lambda",
      "protocol": "HTTP/1.1",
      "sourceIp": "15.248.3.126",
      "userAgent": "HTTPie/3.2.2"
    },
    "requestId": "ZhNHJhhLjoEEPiw=",
    "routeKey": "POST /lambda",
    "stage": "$default",
    "time": "17/Jun/2024:15:52:39 +0000",
    "timeEpoch": 1718639559080
  },
  "isBase64Encoded": false,
  "body": "....." <-- your payload goes here
}
```

With this in mind, you have two options:
1. Use the envelope - like you did 
2. [Extend the `APIGatewayProxyEventV2Schema ` schema](https://docs.powertools.aws.dev/lambda/typescript/latest/utilities/parser/#extend-built-in-schemas)

```ts
const mySchema = APIGatewayProxyEventV2Schema.extend({
  body: newNetsuiteUserSchema
});
```

In both cases Parser will validate both the outer object (aka the envelope - in this case the API Gateway HTTP Proxy event) and the inner payload (aka the body with the schema you provide). The main difference between the two is that the envelope one replaces the `event` and gives you only your data (in this case the parsed `body`) while the other gives you the entire `event` including also all the other fields (i.e. `headers`, `queryString`, etc.).

**Comment by julbrs** [comment]
Hi @dreamorosi,

Thanks for the time to understand my issue. I confirm that I am using a Lambda behind API Gateway v2 so my `event` object contains much more content that only my payload, and the payload is in fact only in the `body` field.

Regarding schema extension, I think it can be a nice solution. I have also tried that but I cannot reproduce the "magic" generated by the Envelope (that is converting the string body to a parsed object!).

Here is an example:

```typescript
import { newNetsuiteUserSchema } from '@/packages/core/model/netsuite';

import middy from '@middy/core';
import { parser } from '@aws-lambda-powertools/parser/middleware';
import { APIGatewayProxyEventV2Schema } from '@aws-lambda-powertools/parser/schemas';
import { z } from 'zod';

// extended APIGatewayProxyEventV2Schema
export const newNetsuiteUserEventSchema = APIGatewayProxyEventV2Schema.extend({
  body: newNetsuiteUserSchema,
});

export const baseHandler = async (event: z.infer<typeof newNetsuiteUserEventSchema>) => {
  console.log(event);
  return {
    statusCode: 200,
    body: JSON.stringify(event),
  };
};

export const handler = middy(baseHandler).use(parser({ schema: newNetsuiteUserEventSchema }));
```

Here when I am calling my API, I end with: 

```
Error: Failed to parse schema. This error was caused by: [
  {
    "code": "invalid_type",
    "expected": "object",
    "received": "string",
    "path": [
      "body"
    ],
    "message": "Expected object, received string"
  }
].
```

This is expected because the body is a stringify version of my payload, not an object ;)

I think I need to provide a custom Envelope here to the `parser` fonction, so the body will be correctly parsed, am I right?

I found the API Gateway v2 Envelope here: https://github.com/aws-powertools/powertools-lambda-typescript/blob/e709c9c24b9e45a5b8850c35f760cdec2fb43c6e/packages/parser/src/envelopes/apigwv2.ts

**Comment by dreamorosi** [comment]
I see what you mean now, you're right we don't do any transform when extending the `body` of a schema.

I was wrong in my previous statement, I mixed the schemas and the envelopes and assumed we were doing the same in both. In my mind, I had the example of `EventBridgeSchema` which we can extend because the `details` is an object itself, but in this case the `body` is a stringified representation of the JSON object.

In this case, the way you'd extend the schema would look something like this:

```ts
export const newNetsuiteUserEventSchema = APIGatewayProxyEventV2Schema.extend({
  body: z
    .string()
    .transform((str, ctx) => {
      try {
        return JSON.parse(str);
      } catch (e) {
        ctx.addIssue({ code: 'custom', message: 'Invalid JSON' });
        return z.NEVER;
      }
    })
    .pipe(newNetsuiteUserSchema),
});
```

I tested the above with the copy/pasted version of your schema and function and the parsing is successful.

I guess the next question would be whether we should add this by default or not. I can see how without it, the schema has somewhat limited use, or at least it'd require a second parsing within the function handler just for the body - assuming one doesn't want to use the envelope, for example because they want to retain access to fields other than the `body`.

I'm going to convert this to a feature request, and regardless of the decision we make on adding it, we should likely consider updating the documentation to include this example.

My main concern here is, how to integrate this in a way that customers can still use the plain `string` if they don't want to specify a schema AND also how to make the `.extend()` cleaner (i.e. without asking people to use `.pipe(mySchema)`.

Would like to loop @am29d here, since he's the Zod expert.

**Comment by julbrs** [comment]
Hi,

Your feedback is highly appreciated! I have opened the issue just 2 hours ago and I have already multiple tracks to follow in order to find the correct way to implement the parser on my own project, thanks a lot!

I will take some time to study your code suggestion, but wow, I don't have this level yet in Zod to handle `transform` stuff, it's new for me! I wasn't aware of this feature!

I am agree that adding a few example like that in the documentation will help new developers that are discovering the project!

Again thanks for your help, it helps me to find the best way to implement the powertools parser!

**Comment by am29d** [comment]
Hey @julbrs ,

thanks for raising the issue and taking the time to share more information so we can better understand and prioritise it. 

I agree that the DX for built-in schemas can be improved without relying on envelopes. The core idea behind `.extend` was to keep the schema simple so customers can fall back to zod functionality and do whatever they need. We don't want to reimplement zod. We already abstract `parse` and `safeParse`, and there needs to be a balance not to overdo it. On the other hand, this is exactly the value we provide by abstracting **common** use cases with parser and zod. 

For ApiGateway schema, I think the majority of use cases will be a json stringified object and it's useful to offer custom `.extend` mechanism to plugin in your schema more easily than using `.transform` and `.pipe` with error handling. 

In addition to Andrea's solution you can also manually parse the `body` of the event as a workaround:

```typescript
import { newNetsuiteUserSchema } from '@/packages/core/model/netsuite';

import middy from '@middy/core';
import { parser } from '@aws-lambda-powertools/parser/middleware';
import { APIGatewayProxyEventV2Schema } from '@aws-lambda-powertools/parser/schemas';
import { z } from 'zod';
import { ApiGatewayV2Envelope } from '@aws-lambda-powertools/parser/envelopes';

export const baseHandler = async (event: ApiGatewayV2Event) => {
  const newNetsuiteUser = newNetsuiteUserSchema.parse(JSON.parse(event.body));
  return {
    statusCode: 200,
    body: JSON.stringify(event),
  };
};

export const handler = middy(baseHandler).use(
  parser({ schema: ApiGatewayV2Envelope })
);
```

Your issue helped us also to identify the gap in the docs. We (unintentionally) picked an easy use case with `EventBridgeSchema` and how to extend it, and will add more examples how to extend other built-in schemas. I am certain this will help us to understand the DX and what rough edges we need polish.

**Comment by julbrs** [comment]
Hi @am29d,

Thanks for the complementary information here!

For now I will end my proof-of-concept using the `.transform()` and `.pipe()` function as it fit perfectly the need here. I feel less confident with the solution of manual parsing inside the handler, because the whole idea of our improvement is to validate the user input outside of the handler, and provide to the handler an object correctly typed. We use to have some challenge to do that on our own with Zod + Middy, and here `aws-powertools/parser` feature is filing the gap I think.

If we can have some optimization to remove the need of providing `.transform()` and `.pipe()`, yes it will be appreciated and we will switch to that!

Again thanks for all the work done around this case, it helps me a lot and I hope it will provide some additional information to other developers that want to dig this nice new tooling you are supporting! 🙏

**Comment by dreamorosi** [comment]
Hi @julbrs, in #2901 we have added a new helper function that should make your use case a bit easier:

```ts
import { JSONStringified } from '@aws-lambda-powertools/parser/helpers';
import { AlbSchema } from '@aws-lambda-powertools/parser/schemas';
import { z } from 'zod';

const customSchema = z.object({
  name: z.string(),
  age: z.number(),
});

const extendedSchema = AlbSchema.extend({
  body: JSONStringified(customSchema),
});
```

This will be available in the next release which should be published this today/tomorrow.

Thank you again for your inputs and for helping improve the project for everyone.

**Comment by julbrs** [comment]
Hi @dreamorosi 

I am testing the new `JSONStringified()` helper, but I have a strange outcome! My IDE doesn't handle correctly the type? 

![image](https://github.com/user-attachments/assets/0a1116c4-1bb3-48ac-a90f-d083bcb8ccc6)

I use to have the correct type by using the workaround described [here](https://github.com/aws-powertools/powertools-lambda-typescript/issues/2844#issuecomment-2250861965). Is it the expected behavior of `JSONStringified()` or do you want me to raise an issue?

Thanks!

**Comment by dreamorosi** [comment]
Hi @julbrs, sorry for the late reply, the team was off due to public holiday in most of EU.

I have checked and this is indeed not how we intended the helper to work. The return type should be inferred from the schema used rather than being a generic `any`.

Upon further troubleshooting I noticed that we made a mistake in how we typed the helper which prevented TypeScript from properly inferring the type.

I have opened an issue to track this #2945 and opened a subsequent PR to fix it #2946. The PR body has a visual confirmation of the fix working.

Unless something major happens, the change should be included in next week's release. 

Thank you for reporting the issue!
